File: 209.长度最小的子数组.cpp
/*
 * @lc app=leetcode.cn id=209 lang=cpp
 *
 * [209] 长度最小的子数组
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    int minSubArrayLen(int target, vector<int> &nums)
    {
        int n = nums.size();
        int start = 0;
        int window_sum = 0;
        int min_len = n + 1;

        for (int end = 0; end < n; ++end)
        {
            window_sum += nums[end];

            while (window_sum >= target)
            {
                min_len = min_len < end - start + 1 ? min_len : end - start + 1;
                window_sum -= nums[start];
                start++;
            }
        }
        return min_len == n + 1 ? 0 : min_len; // 注意不存在最小子数组的情况
    }
};
// @lc code=end

File: 209.长度最小的子数组.py
#
# @lc app=leetcode.cn id=209 lang=python3
#
# [209] 长度最小的子数组
#
from typing import List


# @lc code=start
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        # 0...j,j+1...i

        res = float("inf")

        left = 0
        current_sum = 0
        for right in range(0, len(nums)):
            current_sum += nums[right]
            while current_sum >= target:
                res = min(res, right - left + 1)
                current_sum -= nums[left]
                left += 1

        # 证明没有target
        if res != float("inf"):
            return res
        else:
            return 0


# @lc code=end
sol = Solution()
print(sol.minSubArrayLen(7, [2, 3, 1, 2, 4, 3]))
print(sol.minSubArrayLen(4, [1, 4, 4]))
print(sol.minSubArrayLen(11, [1, 2, 3, 4, 5]))
print(sol.minSubArrayLen(6, [10]))
print(sol.minSubArrayLen(11, [1, 1, 1, 1, 1, 1, 1, 1]))
